### PR TITLE
Support division on ANSI interval types

### PR DESCRIPTION
**Pull request by res-life**
Contributes https://github.com/NVIDIA/spark-rapids/issues/4151

Supports 
- Year-month interval / (byte, short, int, long, float, double)
- Day-time interval / (byte, short, int, long, float, double)

Note: not support decimal type because of CUDF Decimal128 does not handle overflow, like:
```
CUDF positive decimal * positive decimal can generate negative value.
val d1 = ColumnVector.fromDecimals(new java.math.BigDecimal("123456789012345678901234567890123456.11"))
val d2 = Scalar.fromDecimal(new java.math.BigDecimal("12345678901234567890123456789012345611"))
val r = d1.mul(d2)
println(r)

123456789012345678901234567890123456.11 * 12345678901234567890123456789012345611 = -1053265448675579329726886670761974900.55
```

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
Blocked by 2 Spark330 build issues:
https://github.com/NVIDIA/spark-rapids/issues/5123
https://github.com/NVIDIA/spark-rapids/issues/5133

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/Spark33XShims.scala
Comment (jlowe) [inline sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/Spark33XShims.scala:316]: This should leverage the new `TypeSig` interval type shims added in #5020 to avoid copying a lot of code just to override the accepted types for Spark 3.3

**Review by jlowe — COMMENTED** [review]
The temporary changes needed to compile on Spark 3.3 can be removed now that #5124 has been merged.

**Comment by res-life** [comment]
Waiting https://github.com/NVIDIA/spark-rapids/issues/5150 to merge, has some conflict on Spark33XShims.

**Comment by res-life** [comment]
Waiting https://github.com/NVIDIA/spark-rapids/issues/5150 to merge, has some conflict on Spark33XShims.

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/Spark33XShims.scala
Comment (res-life) [reply sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/Spark33XShims.scala]: Updated

**Comment by res-life** [comment]
Rewrited the division code by leveraging `decimal_128` type to reduce overflow check time,  got about 2x speedup.

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala
Comment (jlowe) [inline sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala:198]: While we don't expect to ever hit this, if we do then it will be annoying that the exception didn't bother to state what the unexpected type was.
```suggestion
      case t =>
        throw new IllegalArgumentException(s"Unexpected type $t")
```
Comment applies to the other instances of this below.
Comment applies to the

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala
Comment (jlowe) [inline sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala:289]: Is pre-casting really necessary, or will this happen naturally when we specify the desired output type in the `.div` call below?  If cudf is going to pre-cast for us during the div operation, it would be nice to avoid manifesting the intermediate result explicitly.

Comment (res-life) [reply sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala:289]: Cudf does not do this for us. If so, it will improve the performance.

```
    val dT = DType.create(DType.DTypeEnum.DECIMAL128, -1)
    val cv1 = ColumnVector.fromLongs(5L)
    val cv2 = ColumnVector.fromLongs(3L)
    println(cv1.div(cv2, dT).getScalarElement(0))
```
get Scalar{type=DECIMAL128 scale:-1 **value=1.0**}

But for the `long_cv.mul(long_cv, DECIMAL128(scale = 0))`, we can do it, the only difference is the `scale = 0` instead of `scale = 
 1`

I think we can file a `feature request`  to cuDF, what do you think?

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala
Comment (res-life) [reply sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala]: done

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala
Comment (jlowe) [reply sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala:289]: Yeah I think cudf will only upcast to the "natural" type of the two operands and the operator then cast at the end to the desired type.  Since C++ won't naturally upcast to __int128_t for this operator, I don't think it's reasonable to ask cudf to add this feature.  I'm sure they would just push back and say we need to upcast if we need that behavior.  Otherwise it would add quite a bit of complexity to all their binary operators for arguably a fairly specialized type of behavior.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:132]: Why not use pattern matching (i.e.: `num match { case _: Byte => ...`) here instead of a series of explicit instance checks?  Applies to the other places where it's doing a series of type checks.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:163]: Nit: Should print a more useful message here in case it is hit, otherwise one has to line up the line number in the exception stacktrace with the source to understand what went wrong.  Also recommend using `fail` here since this is a test.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:209]: Nit: use `fail` with a useful error message.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:253]: Nit: use `fail` with a useful error message.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:222]: Nit: pattern matching would be more natural

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:178]: Pattern matching makes more sense here, as it's more concise to write and read and removes the need to cast, e.g.:
```suggestion
        val (data, schema) = num match {
          case n: Byte =>
            (Seq(Row(Period.ofMonths(months), n)),
                StructType(Seq(
                  StructField("c1", YearMonthIntervalType()),
                  StructField("c2", ByteType)
                )))
          case n: Short =>
            ...
```

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

**Comment by res-life** [comment]
build

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:133]: This does not match properly, as it's checking an instance against a class which will never match.  It needs to be written something like this instead:
```suggestion
          case _: Byte => df.selectExpr(s"c1 / cast('$num' as Byte)")
```

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:154]: These cases will never match, needs to be something like this instead:
```suggestion
          case n: Byte => (Seq(Row(Period.ofMonths(months), n)), StructType(Seq(
```

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:193]: Same case matching issue

**Review by jlowe — COMMENTED** [review]
Tests are no longer passing after the recent changes, which implies they were not rerun after the changes were made.  Note that since this test is Spark 3.3.0+, it's not normally run as part of premerge (that may soon change once Spark 3.3 finally ships).  In the meantime, it's important to run it manually or we'll end up finding it doesn't pass on the nightly 3.3.0 runs after it already has been merged.

**Comment by res-life** [comment]
build

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: Sorry for this mistake. 
Updated.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: Updated.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: Updated.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:137]:
```suggestion
          case f: Float if f.isNaN => df.selectExpr("c1 / cast('NaN' as float)")
```

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (jlowe) [inline tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala:143]:
```suggestion
          case d: Double if d.isNaN => df.selectExpr("c1 / cast('NaN' as double)")
```

**Review by jlowe — COMMENTED** [review]
Minor nit on checking for NaN values using the dedicated predicate method, but otherwise lgtm.

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

File: tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala
Comment (res-life) [reply tests/src/test/330/scala/com/nvidia/spark/rapids/IntervalDivisionSuite.scala]: done

**Comment by res-life** [comment]
build

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
CI failure is unrelated, appears to be the (transient?) Cloudera dependency issue.  Kicking the build again.

**Comment by jlowe** [comment]
build

**Pull request merged**
